api/jobs.go

```diff
@@ -779,38 +782,43 @@ func (p *PeriodicConfig) GetLocation() (*time.Location, error) {
 type ParameterizedJobConfig struct {
-	Payload      string
-	MetaRequired []string `mapstructure:"meta_required"`
-	MetaOptional []string `mapstructure:"meta_optional"`
+	Payload      string   `hcl:"payload,optional"`
+	MetaRequired []string `mapstructure:"meta_required" hcl:"meta_required,optional"`
+	MetaOptional []string `mapstructure:"meta_optional" hcl:"meta_optional,optional"`
 }
 
 // Job is used to serialize a job.
 type Job struct {
+	/* Fields parsed from HCL config */
+
+	Region           *string                 `hcl:"region,optional"`
+	Namespace        *string                 `hcl:"namespace,optional"`
+	ID               *string                 `hcl:"id,optional"`
+	Name             *string                 `hcl:"name,optional"`
+	Type             *string                 `hcl:"type,optional"`
+	Priority         *int                    `hcl:"priority,optional"`
+	AllAtOnce        *bool                   `mapstructure:"all_at_once" hcl:"all_at_once,optional"`
+	Datacenters      []string                `hcl:"datacenters,optional"`
+	Constraints      []*Constraint           `hcl:"constraint,block"`
+	Affinities       []*Affinity             `hcl:"affinity,block"`
+	TaskGroups       []*TaskGroup            `hcl:"group,block"`
+	Update           *UpdateStrategy         `hcl:"update,block"`
+	Multiregion      *Multiregion            `hcl:"multiregion,block"`
+	Spreads          []*Spread               `hcl:"spread,block"`
+	Periodic         *PeriodicConfig         `hcl:"periodic,block"`
+	ParameterizedJob *ParameterizedJobConfig `hcl:"parameterized,block"`
+	Reschedule       *ReschedulePolicy       `hcl:"reschedule,block"`
+	Migrate          *MigrateStrategy        `hcl:"migrate,block"`
+	Meta             map[string]string       `hcl:"meta,block"`
+	ConsulToken      *string                 `mapstructure:"consul_token" hcl:"consul_token,optional"`
+	VaultToken       *string                 `mapstructure:"vault_token" hcl:"vault_token,optional"`
+
+	/* Fields set by server, not sourced from job config file */
+
 	Stop              *bool
-	Region            *string
-	Namespace         *string
-	ID                *string
 	ParentID          *string
-	Name              *string
-	Type              *string
-	Priority          *int
-	AllAtOnce         *bool `mapstructure:"all_at_once"`
-	Datacenters       []string
-	Constraints       []*Constraint
-	Affinities        []*Affinity
-	TaskGroups        []*TaskGroup
-	Update            *UpdateStrategy
-	Multiregion       *Multiregion
-	Spreads           []*Spread
-	Periodic          *PeriodicConfig
-	ParameterizedJob  *ParameterizedJobConfig
 	Dispatched        bool
 	Payload           []byte
-	Reschedule        *ReschedulePolicy
-	Migrate           *MigrateStrategy
-	Meta              map[string]string
-	ConsulToken       *string `mapstructure:"consul_token"`
-	VaultToken        *string `mapstructure:"vault_token"`
 	VaultNamespace    *string `mapstructure:"vault_namespace"`
```

Comment:
May require follow up for VaultNamespace. I'm a little surprised that vault_namespace field isn't a valid config option. I checked the field is rejected emperically, and confirmed it's not listed as a valid key in the hclv1 parser: https://github.com/hashicorp/nomad/blob/1d6997942992e92a0eccaec1d87cb0c1d43eeb0b/jobspec/parse_job.go#L79-L82.